### PR TITLE
Hoist collapse shape out of scf.forall when possible and expand its destination

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
This pattern allows to hoist out collapse shape producer of tensor.parallel_insert_slice and expand the enclosing scf.forall destination. This is only safe because we are doing this on workgroup mapped sc.forall's and hence the slices are disjoint. The reason to have this pattern is that it allows us to eliminate empty tensors during bufferization which would otherwise be blocked due to the collapse.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 1]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+  auto destType = dyn_cast<ShapedType>(dest.getType());
+  if (!destType)
+    return failure();
+  // Iterator to insert outer sizes.
+  auto outerShapeIter = expandedShape.begin();
+  for (auto [reassociations, destSize] :
+       llvm::zip_equal(reIndices, destType.getShape())) {
+    int64_t totalInnerSize = 1;
+    for (size_t i = 1, e = reassociations.size(); i < e; ++i) {
+      int64_t expandedInnerSize = sliceStaticSizes[reassociations[i]];
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: Actually, this can be:
```suggestion
    for (int64_t reasociation : llvm::drop_begin(reassociations)) {
      int64_t expandedInnerSize = sliceStaticSizes[reassociation];
```

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 2]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+  }
+  return success();
+}
+
+/// Check if the users of the expanded scf.forall destination can be updated to
+/// account for the expand. If not we bail out. There are two supported users
+/// which are extract_slice -> expand_shape with the same exact reassociation
+/// map as the collapse op to be hoisted out or a parallel_insert_slice.
+static LogicalResult
+verifyandCollectExpandableUsers(Value insertDest,
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]:
```suggestion
verifyAndCollectExpandableUsers(Value insertDest,
```

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 3]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+          affine::makeComposedFoldedAffineApply(
+              rewriter, loc, outerDimExpr,
+              {getValueOrCreateConstantIndexOp(rewriter, loc, offset),
+               rewriter.getIndexAttr(totalInnerSizes[index])}));
+
+      expandedOffsetsIter = expandedOffsets.end();
+    }
+    ArrayRef<int64_t> expandedShape = resultType.getShape();
+    SmallVector<OpFoldResult> expandedSizes;
+    for (auto size : expandedShape) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: nit: not auto here

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 4]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+      expandedSizes.push_back(getAsIndexOpFoldResult(ctx, size));
+    }
+    SmallVector<OpFoldResult> expandedStrides(resultType.getRank(),
+                                              rewriter.getIndexAttr(1));
+    return std::make_tuple(expandedOffsets, expandedSizes, expandedStrides);
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]:
```suggestion
    return {expandedOffsets, expandedSizes, expandedStrides};
```

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 5]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+  };
+  for (auto user : expandableUsers) {
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 6]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+    rewriter.setInsertionPointToStart(forallOp.getBody());
+    if (auto extractSliceOp = dyn_cast<tensor::ExtractSliceOp>(user)) {
+      auto expandShapeOp =
+          dyn_cast<tensor::ExpandShapeOp>(*extractSliceOp->getUsers().begin());
+      auto resultType = expandShapeOp.getResultType();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 7]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+          expandedOffsets, expandedSizes, expandedStrides);
+      rewriter.setInsertionPoint(expandShapeOp);
+      rewriter.replaceOpWithNewOp<tensor::ExpandShapeOp>(
+          expandShapeOp, resultType, expandShapeOp.getSrc(),
+          unFlattenReassociations);
+    } else if (auto parallelInsertOp =
+                   dyn_cast<tensor::ParallelInsertSliceOp>(user)) {
+      auto collapseShapeOp =
+          parallelInsertOp.getSource().getDefiningOp<tensor::CollapseShapeOp>();
+      auto resultType = collapseShapeOp.getSrcType();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: also here

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 8]
```diff
@@ -17,6 +19,273 @@ namespace mlir::iree_compiler {
...
+      return failure();
+    }
+
+    // Expand the users of the destination.
+    rewriter.setInsertionPointToStart(forallOp.getBody());
+    expandVerifiedUsers(rewriter, loc, ctx, expandableUsers, totalInnerSizes,
+                        reIndices, forallOp);
+    rewriter.setInsertionPoint(forallOp);
+
+    auto outOp = forallOutputs[0].getDefiningOp();
```
Comment (kuhar) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: also here

**Review by kuhar — COMMENTED** [review]
Just some nits

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 4 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I tried this one first, it threw a compiler error about not being able to infer the type.

Comment (kuhar) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I'd add casts in this case then, since otherwise you have some implicit conversion over the tuple types instead which seem more scary

Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I got it working by being explicit about the return type of the lambda

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 9]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+    if (!forallOpHasMappingType<IREE::Codegen::WorkgroupMappingAttr>(
+            forallOp)) {
+      return failure();
+    }
+    // This pattern only supports forall ops with single
+    // output.
+    SmallVector<Value> forallOutputs(forallOp.getOutputs());
+    if (forallOutputs.size() != 1) {
+      return failure();
+    }
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I think there is no reason to restrict this to a single output. I imagine the logic should work the same regardless of how many other outputs there are. Could you try some tests with multiple outputs and see if it works?

If the logic is not able to handle multi output, then you don't need to implement it now, but could you leave it as TODO?

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 10]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+                 ArrayRef<int64_t> sliceStaticSizes, Value dest,
+                 SmallVectorImpl<int64_t> &expandedShape,
+                 SmallVectorImpl<int64_t> &totalInnerSizes) {
+  auto destType = dyn_cast<ShapedType>(dest.getType());
+  if (!destType)
+    return failure();
+  // Iterator to insert outer sizes.
+  auto outerShapeIter = expandedShape.begin();
+  for (auto [reassociations, destSize] :
+       llvm::zip_equal(reIndices, destType.getShape())) {
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp:40]: If the parallel_insert_slice is rank-reducing, then I think reassociations and destType can have non equal rank. Could you adda check here that the rank of destType is equal to the number of reassociation lists? (and maybe a TODO to support rank reducing cases)

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 11]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+      int64_t expandedInnerSize = sliceStaticSizes[reasociation];
+      if (ShapedType::isDynamic(expandedInnerSize)) {
+        return failure();
+      }
+      expandedShape.push_back(expandedInnerSize);
+      totalInnerSize *= expandedInnerSize;
+    }
+    if (destSize % totalInnerSize != 0) {
+      return failure();
+    }
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I think can fail when the destSize is dynamic and the totalInnerSize is 1 (i.e., dimension is not collapsed), but that should be allowed, and will probably show up in practice. Can you add a special case for when `reassociations` has a size of 1?

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 12]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+      if (reIndices != expandReIndices) {
+        return failure();
+      }
+      expandableUsers.push_back(user);
+    } else if (auto parallelInsertOp =
+                   dyn_cast<tensor::ParallelInsertSliceOp>(user)) {
+      expandableUsers.push_back(user);
+    } else {
+      return failure();
+    }
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: nit: can save a level of nesting on the long extract_slice logic.
```suggestion
    auto parallelInsertOp = dyn_cast<tensor::ParallelInsertSliceOp>(user);
    if (parallelInsertOp) {
      expandableUsers.push_back(user);
      continue;
    }
    auto extractSliceOp = dyn_cast<tensor::ExtractSliceOp>(user);
    if (!extractSliceOp) {
      return failure();
    }
    auto expandShapeOp =
        dyn_cast<tensor::ExpandShapeOp>(*extractSliceOp->getUsers().begin());
    if (!expandShapeOp) {
      return failure();
    }
    SmallVector<ReassociationIndices> expandReIndices =
        expandShapeOp.getReassociationIndices();
    if (reIndices != expandReIndices) {
      return failure();
    }
    expandableUsers.push_back(user);
```

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 13]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+    outerShapeIter = expandedShape.end();
+  }
+  return success();
+}
+
+/// Check if the users of the expanded scf.forall destination can be updated to
+/// account for the expand. If not we bail out. There are two supported users
+/// which are extract_slice -> expand_shape with the same exact reassociation
+/// map as the collapse op to be hoisted out or a parallel_insert_slice.
+static LogicalResult
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I think it is safest to also check that the offsets and sizes are the same as the root parallel_insert_slice op for any extract_slice ops. You can simply check that the OpFoldResults are equal, which will compare exact ssa values for dynamic offsets. This is what happens in bufferization too, so it makes sense to do it here.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 14]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+      RankedTensorType resultType = expandShapeOp.getResultType();
+      auto [expandedOffsets, expandedSizes, expandedStrides] =
+          computeExpandedAccess(extractSliceOp.getMixedOffsets(), resultType);
+      rewriter.replaceOpWithNewOp<tensor::ExtractSliceOp>(
+          extractSliceOp, resultType, extractSliceOp.getSource(),
+          expandedOffsets, expandedSizes, expandedStrides);
+      rewriter.setInsertionPoint(expandShapeOp);
+      rewriter.replaceOpWithNewOp<tensor::ExpandShapeOp>(
+          expandShapeOp, resultType, expandShapeOp.getSrc(),
+          unFlattenReassociations);
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I don't think you need this expand_shape. It is just a nop right?

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 15]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+      auto collapseShapeOp =
+          parallelInsertOp.getSource().getDefiningOp<tensor::CollapseShapeOp>();
+      RankedTensorType resultType = collapseShapeOp.getSrcType();
+      auto [expandedOffsets, expandedSizes, expandedStrides] =
+          computeExpandedAccess(parallelInsertOp.getMixedOffsets(), resultType);
+
+      rewriter.setInsertionPoint(collapseShapeOp);
+      auto newCollapseOp = rewriter.replaceOpWithNewOp<tensor::CollapseShapeOp>(
+          collapseShapeOp, collapseShapeOp.getSrcType(),
+          collapseShapeOp.getSrc(), unFlattenReassociations);
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: Same here, I don't think you need the collapse_shape.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 16]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+
+    Operation *outOp = forallOutputs[0].getDefiningOp();
+    if (!outOp) {
+      return failure();
+    }
+
+    // Create the expand -> new scf.forall -> collapse chain.
+    Type expandedDestType = RankedTensorType::get(
+        expandedDestShape,
+        cast<ShapedType>(outOp->getResult(0).getType()).getElementType());
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: nit: You can use clone to compute the type.
```suggestion
    auto expandedDestType = cast<RankedTensorType>(outOp->getResult(0).getType()).clone(expandedDestShape);
```

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 17]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+    // Expand the users of the destination.
+    rewriter.setInsertionPointToStart(forallOp.getBody());
+    expandVerifiedUsers(rewriter, loc, ctx, expandableUsers, totalInnerSizes,
+                        reIndices, forallOp);
+    rewriter.setInsertionPoint(forallOp);
+
+    Operation *outOp = forallOutputs[0].getDefiningOp();
+    if (!outOp) {
+      return failure();
+    }
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: Why does this need to check for a producer operation? Can you just use `forallOutputs[0]` directly?

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp [hunk 18]
```diff
@@ -17,6 +19,275 @@ namespace mlir::iree_compiler {
...
+    auto collapsedResultOp = rewriter.create<tensor::CollapseShapeOp>(
+        loc, cast<ShapedType>(forallOp->getResult(0).getType()),
+        newForallOp->getResult(0), reIndices);
+
+    // Merge the old scf.forall block which has the expanded users into the new
+    // scf.forall which has the expanded destination.
+    SmallVector<Value> argReplacements(newForallOp.getInductionVars());
+    for (auto forallIterArg : newForallOp.getRegionIterArgs()) {
+      argReplacements.push_back(forallIterArg);
+    }
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: nit: I think this can be
```
argReplacements.append(newForallOp.getRegionIterArgs().begin(),
                       newForallOp.getRegionIterArgs().end());
```

File: compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir [hunk 19]
```diff
@@ -86,3 +87,103 @@ func.func @fold_collapse_into_stores_dynamic(%arg0 : tensor<2x?x32xf32>) {
 //       CHECK:   flow.dispatch.tensor.store %{{.+}}, %[[SUBSPAN]]
 //  CHECK-SAME:       offsets = [0, 0, 0], sizes = [2, %[[SHAPE]], 32], strides = [1, 1, 1]
 //  CHECK-SAME:       !flow.dispatch.tensor<writeonly:tensor<2x?x32xf32>>{%[[SHAPE]]}
+
+// -----
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir:91]: Can you also add a test with dynamic dims? Maybe one with a dynamic dim blocking expansion, and one with mixed dynamic and static dims that can be expanded?

Maybe also something that tests the conditions in `verifyAndCollectExpandableUsers`.

File: compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir [hunk 20]
```diff
@@ -86,3 +87,103 @@ func.func @fold_collapse_into_stores_dynamic(%arg0 : tensor<2x?x32xf32>) {
...
+    %expanded = tensor.expand_shape %extracted_slice [[0, 1], [2, 3, 4]]
+              output_shape [8, 16, 8, 8, 2] : tensor<128x128xf32> into tensor<8x16x8x8x2xf32>
+    %6 = linalg.copy ins(%transposed : tensor<8x16x8x8x2xf32>)
+         outs(%expanded : tensor<8x16x8x8x2xf32>) -> tensor<8x16x8x8x2xf32>
+    %collapsed = tensor.collapse_shape %6 [[0, 1], [2, 3, 4]] : tensor<8x16x8x8x2xf32> into tensor<128x128xf32>
+    scf.forall.in_parallel {
+      tensor.parallel_insert_slice %collapsed into %arg2[%arg0, %arg1] [128, 128] [1, 1]
+        : tensor<128x128xf32> into tensor<2048x10240xf32>
+    }
+  } {mapping = [#iree_codegen.workgroup_mapping<y>, #iree_codegen.workgroup_mapping<x>]}
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir]: This is a lengthy test. Can you shorten the test by removing all the inner operations except the ones that are necessary for the test?

**Review by Max191 — CHANGES_REQUESTED** [review]
So far looking good! I left some comments about some potential edge cases, and some more tests that would be good to have for the first round of comments.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 11 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: we dont want to fail for totalInnerSize=1, as you mentioned that is very common for unexdpanded dims and is supported already, the logic will just continue with the next dim. I will add a test for that. 

Will add the check for dynamic destSize.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 14 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I tried replacing users with new extractslice and just erasing it but that was leading to segfaults in future patterns, saw this issue in  three cases
1. this expand
2. the collapse
3. old scf.forall
The IR after the pattern looked good but then other patterns still tried to apply on the erased ops, so i found it was easier to just make them no-ops or leave them with no users and let the canonicalizations do their thing.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 15 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: Answered on the expand.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 9 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: I can add it, just will have to keep track of which output I am applying the pattern on then, as it needs to be expanded.

File: compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir (hunk 19 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir:91]: Can definitely add the test but will have to see if its easy to actually do the expansion. Usually in dynamic case there is an extract_slice after the collapse so this pattern cant be applied anyway so I dont see much value in supporting the dynamic case.

File: compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir (hunk 20 above)
Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir]: I tried to make the shortest possible test but even removing one of the inner ops causes other patterns/canonicalizations to kick in such that the tested pattern is not even needed.

File: compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp (hunk 14 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: > I tried replacing users with new extractslice and just erasing it

If you were explicitly erasing the op in this function, then that might have been what was causing it. Usually it is best not to completely erase ops that are not the root op of the pattern, since other patterns can end up trying to access those ops if they were already on the worklist. Does it work if you replace the users of the op but don't erase it?

Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/PropagateReshapesByExpansion.cpp]: Actually was just trying this and you are exactly right, if I dont erase then it works, thanks for catching this reduced a lot of unnecessary code.

File: compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir (hunk 19 above)
Comment (Max191) [reply compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir:91]: I see, yeah I think it makes sense for packed dynamic dimensions. I was also imagining a case like a batch matmul with a dynamic batch dimension. The static M/N/K dimensions would be unpacked/collapsed, but the outer batch dimension is untouched. I think we want that to still work for this pattern right?

Comment (nirvedhmeshram) [reply compiler/src/iree/compiler/Codegen/Common/test/propagate_reshapes_by_expansion.mlir:91]: Ya that makes sense, so we basically want to support dynamic destination sizes then, is that right?

**Review by Max191 — APPROVED** [review]
Looks good! The code looks a lot cleaner now too, nice work!

**Comment by Max191** [comment]
Hmm, looks like punet fails to compile with this. I think it must be one of the conv dispatches (because nothing else uses this pass), but it seems a bit strange to me that this is causing a compilation failure.

**Comment by nirvedhmeshram** [comment]
> Hmm, looks like punet fails to compile with this. I think it must be one of the conv dispatches (because nothing else uses this pass), but it seems a bit strange to me that this is causing a compilation failure.

Yeah thats a bummer, one thing I can try is to see if CI passes when I disable multi-result, if thats the case then the issue might be that the getTiedResult function assumes the multiple parallel_insert_slice are ordered in the same way as the results.

**Comment by Max191** [comment]
I took a look at the model, and found the dispatch that is failing. Here's a gist with a dump: https://gist.github.com/Max191/a092ec5c81e5f9f44aee19358f007a34

Looks like the problem is that the collapse_shape was hoisted out of the forall, but then could not fold with the flow.dispatch.tensor.store at the function boundary, because the store has an offset (from pad fusion). I'm not quite sure what the best solution for this is at the moment.

EDIT: snippet from the dump after reshape propagation:
```mlir
    } -> (tensor<1x2x8x16x4x16xf16>, tensor<1x2x8x16x4x16xi8>)
    scf.forall.in_parallel {
      tensor.parallel_insert_slice %27#0 into %arg3[%arg0, %arg1, 0, 0, %arg2, 0] [1, 2, 8, 16, 4, 16] [1, 1, 1, 1, 1, 1] : tensor<1x2x8x16x4x16xf16> into tensor<2x128x8x16x20x16xf16>
      tensor.parallel_insert_slice %27#1 into %arg4[%arg0, %arg1, 0, 0, %arg2, 0] [1, 2, 8, 16, 4, 16] [1, 1, 1, 1, 1, 1] : tensor<1x2x8x16x4x16xi8> into tensor<2x128x8x16x20x16xi8>
    }
  } {mapping = [#iree_codegen.workgroup_mapping<z>, #iree_codegen.workgroup_mapping<y>, #iree_codegen.workgroup_mapping<x>]}
  %collapsed = tensor.collapse_shape %20#1 [[0], [1], [2, 3], [4, 5]] : tensor<2x128x8x16x20x16xi8> into tensor<2x128x128x320xi8>
  flow.dispatch.tensor.store %collapsed, %7, offsets = [0, 1, 1, 0], sizes = [2, 128, 128, 320], strides = [1, 1, 1, 1] : tensor<2x128x128x320xi8> -> !flow.dispatch.tensor<readwrite:tensor<2x130x130x320xi8>>
  flow.dispatch.tensor.store %20#0, %8, offsets = [0, 0, 0, 0, 0, 0], sizes = [2, 128, 8, 16, 20, 16], strides = [1, 1, 1, 1, 1, 1] : tensor<2x128x8x16x20x16xf16> -> !flow.dispatch.tensor<writeonly:tensor<2x128x8x16x20x16xf16>>
  return
}
```

**Comment by Max191** [comment]
One way to fix this would be to check that the consumer of the scf.forall is a flow.dispatch.tensor.store with a full slice (i.e., zero offsets, unit strides, and source_sizes == target_sizes). This is making the pattern even more specific, but it is already quite specific, so maybe it is okay for now.

**Comment by nirvedhmeshram** [comment]
> One way to fix this would be to check that the consumer of the scf.forall is a flow.dispatch.tensor.store with a full slice (i.e., zero offsets, unit strides, and source_sizes == target_sizes). This is making the pattern even more specific, but it is already quite specific, so maybe it is okay for now.

Makes sense, I guess theoretically it should be possible to support this?
```
flow.dispatch.tensor.store %20#1, %7, offsets = [0, 1, 0, 1, 0, 0], sizes = [2, 128, 8, 16, 20, 16], strides = [1, 1, 1, 1, 1, 1] : tensor<2x128x8x16x20x16xf16> -> !flow.dispatch.tensor<writeonly:tensor<2x130x130x20x16xf16>
```

**Comment by Max191** [comment]
Following up here from a discussion with Nirvedh offline. We are going with the solution I proposed here: https://github.com/iree-org/iree/pull/19044#issuecomment-2470913415

As long as the convolution has a consumer in the dispatch, then the pipeline should be able to handle not hoisting the collapse_shape when the dispatch.tensor.store has offsets (when a pad consumer has been fused). If there is no consumer, and the result of the convolution is directly stored with some offsets, then this will cause problems, but I don't think it is a case that is likely to be seen. There is typically an operation between convolutions and pads, and pad fusion does not happen by default anyway. It is a case to be aware of, but I think it is fine to proceed with this additional restriction for now.

In other words, the following case would be problematic if the pad is fused with the producer convolution:
```mlir
%conv = linalg.conv_2d_nhwc_hwcf ...
%pad = tensor.pad %conv ...
```
However, this seems like an uncommon case, and pad fusion only happens with aggressive fusion anyway.

File: compiler/src/iree/compiler/Codegen/Utils/Utils.cpp [hunk 21]
```diff
@@ -1511,4 +1511,30 @@ computeDimUpperBound(Value shapedValue, unsigned dimNum,
   return roundedDimBound.getSize();
 }
 
+static bool isAllConstantValue(ArrayRef<OpFoldResult> ofrs, int64_t v) {
+  return llvm::all_of(
+      ofrs, [&](OpFoldResult ofr) { return isConstantIntValue(ofr, v); });
+}
+
```
Comment (Max191) [inline compiler/src/iree/compiler/Codegen/Utils/Utils.cpp]: nit: I recently added this upstream as `areAllConstantIntValue`, so we don't need this function anymore.

**Review by Max191 — APPROVED** [review]
